### PR TITLE
Renderer improvements

### PR DESCRIPTION
**Pull request by tschw**
Contains performance optimizations, minor fixes, WebGL program introspection and GLSL error feedback.

### CODE REVIEW

**Comment by mrdoob** [comment]
I think I would start by making `info` private...

**Comment by tschw** [comment]
Not sure I get what you mean. 

Isn't it for client code to get feedback about resource usage?

```javascript
var _info = { /* ... deeply nested stuff */ };
```
still means a lot of property lookup.

Or one local private (=== upvalue) variable for each member inside that structure and a corresponding accessor? That would be rather bloaty.

I'm puzzled.

**Comment by tschw** [comment]
Constructor:
```javascript
var _infoEnabled = false;
var _info = {

    // [...]

};

// [...]

this.setInfoEnabled = function(enable) {

    _infoEnabled = enable;

};

this.getInfo = function() {

    return _info;

};

// [...]

// DEPRECATED

// [...]

Object.defineProperties( this, {

   // [...]

    info: {
        get: function() {
            console.warn( '...' );
            _infoEnabled = true;
            return _info;
        }
    }
});
```

Rendering code:
```javascript
if (_infoEnabled) {

    // [... update values]

}
```

Did you mean something like that?
Or does it look likable to you?

It would also allow for the whole enchilada to get DCEd when `setInfoEnabled` never gets called - simple enough to be within the reach of JS-based minifiers like Uglify.js.

**Comment by mrdoob** [comment]
Yes, I would start by doing just this:

```javascript

var info = {}

Object.defineProperties( this, {
    info: {
        get: function () {
            return info;
        }
    }
});
```

That way we can replace all the `this.info` for `info`.

**Comment by tschw** [comment]
Awe! But `_this.info` is gone in this changeset already, and without `defineProperties` oddness... This means you did't like it? Or you didn't see it?

The reason why I did what I did and the consitency I had in mind should become apparent looking at the most recent commit that implements a way to introspect the `WebGLProgram` instances produced by the renderer.

**Comment by mrdoob** [comment]
> Awe! But `_this.info` is gone in this changeset already, and without `defineProperties` oddness... This means you did't like it? Or you didn't see it?

Oh, I saw it. But you also changed the whole `info` structure, which breaks old code... Oh! Wait no, you didn't. Sorry 😅

**Comment by mrdoob** [comment]
Ok, could you rename  `_memoryInfo` and `_renderInfo` to `infoMemory` and `infoRender`?

File: src/renderers/webgl/WebGLProgram.js [hunk 1]
```diff
@@ -379,22 +379,52 @@ THREE.WebGLProgram = ( function () {
...
+			console.warn( 'THREE.WebGLProgram: gl.getProgramInfoLog()', programLog );
 
-			console.warn( 'THREE.WebGLProgram: gl.getProgramInfoLog()', programLogInfo );
+		} else if ( vertexLog === '' || fragmentLog === '' ) {
+
+			haveDiagnostics = false;
 
 		}
 
+		if ( haveDiagnostics ) this.diagnostics = {
```
Comment (mrdoob) [inline src/renderers/webgl/WebGLProgram.js]: This is a bit scary code... Can you change it to:

```javascript
if ( haveDiagnostics ) {

    this. diagnostics = { ... }

}
```

**Comment by tschw** [comment]
Quite a little zoo. The last five commits are new + 2 for the renaming / formatting requested above.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**